### PR TITLE
Add more basic types

### PR DESCRIPTION
**Pull request by jon-betts**
Adds support for:

* `Any.tuple`
* `Any.float`
* `Any.complex`
* `Any.generator`

Also resolves some issues with generators and adds a little test coverage.

### CODE REVIEW

File: src/h_matchers/interface.py [hunk 1]
```diff
@@ -26,17 +19,23 @@ class Any(AnyThing):
...
-    set = AnySet
-    dict = AnyDict
+    mapping = collection.AnyMapping
+    dict = collection.AnyDict
+
+    iterable = collection.AnyCollection
+    list = collection.AnyList
+    set = collection.AnySet
+    tuple = collection.AnyTuple
+    generator = collection.AnyGenerator
```
Comment (jon-betts) [inline src/h_matchers/interface.py:38]: Adding more types to the interface, and grouping them a bit more sensibly

File: src/h_matchers/matcher/collection/_mixin/contains.py [hunk 2]
```diff
@@ -88,7 +89,11 @@ def _check_contains(self, other, original=None):
         else:
             matcher_class = AnyIterableWithItems
 
-        if matcher_class(self._items) != original:
+        # If the original object was a generator compare against our copy of
+        # the values, not the original object, as it will have been consumed
+        compare_to = other if isinstance(original, GeneratorType) else original
+
+        if matcher_class(self._items) != compare_to:
```
Comment (jon-betts) [inline src/h_matchers/matcher/collection/_mixin/contains.py:96]: This is the bug that stopped generator comparison working right from `AnyCollection`

File: tests/unit/data_types.py [hunk 3]
```diff
@@ -78,6 +85,10 @@ class Groups:
     STRINGS = {DataTypes.STRING, DataTypes.FALSY_STRING}
 
     INTS = {DataTypes.INT, DataTypes.FALSY_INT}
+    FLOATS = {DataTypes.FLOAT, DataTypes.FALSY_FLOAT}
+    COMPLEX = {DataTypes.COMPLEX, DataTypes.FALSY_COMPLEX}
+    DECIMAL = {DataTypes.DECIMAL, DataTypes.FALSY_DECIMAL}
+    NUMERIC = INTS | FLOATS | COMPLEX | DECIMAL
```
Comment (jon-betts) [inline tests/unit/data_types.py:91]: We don't use all these classes in the tests yet, but we might want to soon

File: tests/unit/h_matchers/matcher/collection/containment_test.py [hunk 4]
```diff
@@ -93,11 +93,17 @@ def test_it_matches_out_of_order(self):
     def test_it_matches_generators_out_of_order(self):
         matcher = AnyIterableWithItems([2, 0, 1])
 
-        assert matcher == iter(range(3))
-        assert iter(range(3)) == matcher
+        def matching_gen():
+            yield from range(3)
```
Comment (jon-betts) [inline tests/unit/h_matchers/matcher/collection/containment_test.py:97]: I'm not 100% sure `iter(range(...))` actually gives you a generator. I think you might get a weird internal Python thing. This definitely does.

File: tests/unit/h_matchers/matcher/collection_test.py [hunk 5]
```diff
@@ -96,4 +118,33 @@ def items(self):
...
         assert TestObject() == AnyMapping()
         assert dict() == AnyMapping()
-        assert list() != AnyMapping()
+
+    @pytest.mark.parametrize("non_matching", (tuple(), list(), set()))
+    def test_non_matching_items(self, non_matching):
+        assert non_matching != AnyMapping()
+
+
+class TestTypeSpecificCollectionMatchers:
```
Comment (jon-betts) [inline tests/unit/h_matchers/matcher/collection_test.py:127]: Test all of these in one go. They don't have much functionality of their own

**Review by seanh — APPROVED** [review]

**Pull request merged**